src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/HelperMatcher.cs

```diff
@@ -0,0 +1,170 @@
+            while (line is not null)
+            {
+                // Helper declarations need to be on a single line, so it's safe
+                // to evaluate the regex on individual lines.
+                if (HelperDeclRegex.IsMatch(line))
+                {
+                    _logger.LogInformation("Razor document {FilePath} contains a @helper function", document.Source.FilePath);
+                    return true;
+                }
+
+                line = await reader.ReadLineAsync().ConfigureAwait(false);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Creates HelperReplacements for each @helper function in a Razor document.
+        /// </summary>
+        /// <param name="document">The Razor document to find @helper functions in.</param>
+        /// <returns>HelperReplacements proposing replacements of each @helper with a local method.</returns>
+        public async Task<IEnumerable<HelperReplacement>> GetHelperReplacementsAsync(RazorCodeDocument document)
+        {
+            if (document is null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            // Unfortunately, there's no good way to walk or update Razor syntax trees, so
+            // the Razor document contents are primarily treated as a string.
+            var path = document.Source.FilePath;
+            using var reader = new StreamReader(path);
+            var text = await reader.ReadToEndAsync().ConfigureAwait(false);
+
+            if (text is null)
+            {
+                _logger.LogWarning("Failed to read Razor document {FilePath}", path);
+                return Enumerable.Empty<HelperReplacement>();
+            }
+
+            var helpers = HelperDeclRegex.Matches(text);
+
+            // This would be a lot more succinct with Linq,
+            // but MatchCollection doesn't implement IEnumerable<Match>.
+            var replacements = new HelperReplacement[helpers.Count];
+            for (var i = 0; i < helpers.Count; i++)
+            {
+                replacements[i] = GetHelperReplacement(text, helpers[i].Index);
+            }
+
+            return replacements;
+        }
+
+        private static HelperReplacement GetHelperReplacement(string text, int helperOffset)
+        {
+            var newText = new StringBuilder("@{ HelperResult ");
+            var functionBodyBegun = false;
+            var braceCount = 0;
+            var index = helperOffset + "@helper ".Length;
+            var lineStart = index;
+
+            // Copy the helper function until the end of its body block
+            while ((!functionBodyBegun || braceCount > 0) && index < text.Length)
+            {
+                var nextChar = text[index];
+
+                // Match braces while moving through the function body so that
+                // the loop can end when all braces are matched
+                braceCount += nextChar switch
```

Comment:
What about a lone brace in a string?

Reply:
Hm... that's a good point. This is where working with the Razor syntax tree would come in handy, but the .NET Core Razor engine doesn't understand @helper syntax so it isn't able to properly evaluate this code block. I may leave this as it is for now knowing that there are some corner cases it won't handle well and we can re-visit if they become problematic in practice.

Possible solutions would range from counting `'` and `"`, too, to give a best-effort attempt at only match real {} to (if we're really serious) creating Razor engine extensions to teach it about @helper syntax so it can parse this properly.